feat: Provides 'Adding a Visual Studio Code Extension' module (#2249)
codenvy/che-docs: modules/user-guide/pages/adding-a-vscode-extension.adoc

:_content-type: assembly
:description: Adding a Visual Studio Code Extension
:keywords: Visual Studio Code extension, user-guide
:navtitle: Adding a Visual Studio Code Extension
// :page-aliases:

[id="adding-visual-studio-code-extension_{context}"]
= Adding a Visual Studio Code Extension

IDE specific plug-ins are no longer part of devfile with devfile v2. Instead, it uses the specific meta-folder to describe the plug-ins.

[id="visual-studio-code-extensions-json"]
== Using `.vscode/extensions.json`

The easiest way to add a Visual Studio Code extension is to add the extension to the `.vscode/extensions.json` file. Using Che-Theia the extension will be automatically installed and configured. With Che-Code, the Visual Studio Code fork, at the startup of the workspace, Visual Studio Code shows a recommendation pop-up to install the extension.

.Procedure

> . Add `.vscode/extensions.json` file if missing at the root of the Git repository

Reviewer: ```suggestion
1. Add `.vscode/extensions.json` file if it is missing at the root of the Git repository.
```

> . Add the extension using its identifier. Publisher and extension name are using a dot separated format.

Reviewer: ```suggestion
1. Add the extension using its identifier. Publisher and extension names use a dot-separated format.
```

> . Example adding the Visual Studio Code Java extension:

Reviewer: ```suggestion
1. Example of adding the Visual Studio Code Java extension:
```

```json
{
  "recommendations": [
    "redhat.java"
  ]
}
```

> . {prod-short} will automatically install Red Hat Visual Studio Code Java extension in the {prod-short} workspace.

Reviewer: ```suggestion
4. {prod-short} will automatically install the Red Hat Visual Studio Code Java extension in the {prod-short} workspace.
```

> Note: if the specified set of extensions IDs are not available in the {prod-short} registry, the workspace creation will not fail but the extensions will not be available.

Reviewer: ```suggestion
Note: If the specified set of extensions IDs is not available in the {prod-short} registry, the workspace creation will not fail but the extensions will not be available.
```

The main advantage is that it will work even when not using {prod-short} as primary IDE.

[id="che-theia-plug-ins-YAML"]
== Using `.che/che-theia-plugins.yaml`

Relying on `.vscode/extensions.json` is simple but sometimes it requires to specify some extra parameters for extensions.

Here comes the optional `.che/che-theia-plugins.yaml` file definition

=== Defines the extensions to install in the workspace

.Procedure

> . Add `.che/che-theia-plugins.yaml` file if missing at the root of the Git repository

Reviewer: ```suggestion
1. Add `.che/che-theia-plugins.yaml` file if it is missing at the root of the Git repository.
```

> . Add the id of the extension using the / separated format

Reviewer: ```suggestion
1. Add the id of the extension using the '/' separated format.
```

```YAML
- id: redhat/java  
```

=== Change default memory limit

> For example, overriding some container settings, for example the memoryLimit for the RedHat Visual Studio Code Java extension.

Reviewer: ```suggestion
For example, override some container settings, like the memory limit for the RedHat Visual Studio Code Java extension.
```

.Procedure

> . Add `override` section in the `.che/che-theia-plugins.yaml` file under the `id` of the extension

Reviewer: ```suggestion
1. Add `override` section in the `.che/che-theia-plugins.yaml` file under the `id` of the extension.
```

> . Specify the memory limit for the extension

Reviewer: ```suggestion
1. Specify the memory limit for the extension.
```

```YAML
- id: redhat/java  
  override:  
    sidecar:  
      memoryLimit: 2Gi
```

> . {prod-short} will automatically install Red Hat Visual Studio Code Java extension in the {prod-short} workspace and increase the memory of the workspace by two gigabytes.

Reviewer: ```suggestion
3. {prod-short} will automatically install the Red Hat Visual Studio Code Java extension in the {prod-short} workspace and increase the memory of the workspace by two gigabytes.
```

=== Override default preferences

> Override the default preferences of the Visual Studio Code extension for the workspace

Reviewer: ```suggestion
Override the default preferences of the Visual Studio Code extension for the workspace.
```

.Procedure

> . Add `override` section in the `.che/che-theia-plugins.yaml` file under the `id` of the extension

Reviewer: ```suggestion
1. Add `override` section in the `.che/che-theia-plugins.yaml` file under the `id` of the extension.
```

> . Specify the preferences into `preferences` section

Reviewer: ```suggestion
1. Specify the preferences in the `Preferences` section.
```

```YAML
- id: redhat/java  
  override:  
    preferences:
      java.server.launchMode: LightWeight
```

> . {prod-short} will automatically install Red Hat Visual Studio Code Java extension in the {prod-short} workspace and sets the `java.server.launchMode` preference to `LightWeight`.

Reviewer: ```suggestion
3. {prod-short} will automatically install Red Hat Visual Studio Code Java extension in the {prod-short} workspace and sets the `java.server.launchMode` preference to `LightWeight`.
```

This method allows to override the default preferences or memory limit of the Visual Studio Code extension for the workspace.

Note: It is also possible to define the preferences in the `.vscode/settings.json` file.

[id="visual-studio-code-extensions-in-devfile"]
== Defines Visual Studio Code extensions in the devfile

The devfile v2 specification is not allowing to define the Visual Studio Code extensions in the devfile. But it is possible to define some attributes in the devfile.

This definition is helpful when it is not possible to add extra files in the repository but is not the recommended way.

Defines the Visual Studio Code extensions by inlining them in the devfile.

=== Inlining `.vscode/extensions.json` file

.Procedure

> . Edit `devfile.yaml` file and add if missing the `attributes` section at the root of the devfile

Reviewer: ```suggestion
1. Edit the `devfile.yaml` file and add if it is missing the `attributes` section at the root of the devfile.
```
